pkg/scheduler/util/predicate_helper.go

```diff
@@ -0,0 +1,108 @@
+package util
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog"
+	"volcano.sh/volcano/pkg/scheduler/api"
+)
+
+type PredicateHelper interface {
+	PredicateNodes(task *api.TaskInfo, nodes []*api.NodeInfo, fn api.PredicateFn) ([]*api.NodeInfo, *api.FitErrors)
+}
+
+type predicateHelper struct {
+	taskPredicateErrorCache map[string]map[string]error
+}
+
+// PredicateNodes returns the specified number of nodes that fit a task
+func (ph *predicateHelper) PredicateNodes(task *api.TaskInfo, nodes []*api.NodeInfo, fn api.PredicateFn) ([]*api.NodeInfo, *api.FitErrors) {
+	var errorLock sync.RWMutex
+	fe := api.NewFitErrors()
+
+	allNodes := len(nodes)
+	if allNodes == 0 {
+		return make([]*api.NodeInfo, 0), fe
+	}
+	numNodesToFind := CalculateNumOfFeasibleNodesToFind(int32(allNodes))
+
+	//allocate enough space to avoid growing it
+	predicateNodes := make([]*api.NodeInfo, numNodesToFind)
+
+	numFoundNodes := int32(0)
+	processedNodes := int32(0)
+
+	taskGroupid := taskGroupID(task)
+	nodeErrorCache, taskFailedBefore := ph.taskPredicateErrorCache[taskGroupid]
+	if nodeErrorCache == nil {
+		nodeErrorCache = map[string]error{}
+	}
+
+	//create a context with cancellation
+	ctx, cancel := context.WithCancel(context.Background())
+
+	checkNode := func(index int) {
+		// Check the nodes starting from where is left off in the previous scheduling cycle,
+		// to make sure all nodes have the same chance of being examined across pods.
+		node := nodes[(lastProcessedNodeIndex+index)%allNodes]
+		atomic.AddInt32(&processedNodes, 1)
+		klog.V(4).Infof("Considering Task <%v/%v> on node <%v>: <%v> vs. <%v>",
+			task.Namespace, task.Name, node.Name, task.Resreq, node.Idle)
+
+		// Check if the task had "predicate" failure before.
+		// And then check if the task failed to predict on this node before.
+		if taskFailedBefore {
```

Comment:
But node resource will be changed anytime, right?
so l think that we cannot assume current task will fail to predicate based on historical failure.

Reply:
Yes, node resource will change anytime. Thus I record the failure in every scheduling perioid, not the entire volcano lifetime. 
If the replica is from same task, its "predicate" failure could be reuse for other replicas. And no need to go predicating for  other replicas in this scheduling. I think this could help improve throughput.

Reply:
yeah l mean in one scheduling period, node resource may change a lot.
e.g when you fail to predicate for task 1 on node, but some pods released on this node immediately, the next task2 should be scheduled to that node successfully.

Reply:
ok, got you now.